core/node/l1_recovery/src/l1_fetcher/l1_fetcher.rs

```diff
@@ -0,0 +1,1083 @@
+                err_code == Some(-32603)             // Internal error
+                    || err_message.contains("failed")    // Server error
+                    || err_message.contains("timed out") // Time-out error
+            };
+
+            // check whether the error is related to having too many results
+            if err_message.contains(TOO_MANY_RESULTS_INFURA)
+                || err_message.contains(TOO_MANY_RESULTS_ALCHEMY)
+                || err_message.contains(TOO_MANY_RESULTS_RETH)
+                || err_message.contains(TOO_BIG_RANGE_RETH)
+                || err_message.contains(TOO_MANY_RESULTS_CHAINSTACK)
+            {
+                // get the numeric block ids
+                let from_number = match from {
+                    BlockNumber::Number(num) => num,
+                    _ => {
+                        // invalid variant
+                        return result;
+                    }
+                };
+                let to_number = match to {
+                    BlockNumber::Number(num) => num,
+                    BlockNumber::Latest => self.eth_client.block_number().await?,
+                    _ => {
+                        // invalid variant
+                        return result;
+                    }
+                };
+
+                // divide range into two halves and recursively fetch them
+                let mid = (from_number + to_number) / 2;
+
+                // safety check to prevent infinite recursion (quite unlikely)
+                if from_number >= mid {
+                    tracing::warn!("Infinite recursion detected while getting events: from_number={from_number:?}, mid={mid:?}");
+                    return result;
+                }
+
+                tracing::warn!("Splitting block range in half: {from:?} - {mid:?} - {to:?}");
+                let mut first_half = self
+                    .get_events_inner(
+                        from,
+                        BlockNumber::Number(mid),
+                        topics1.clone(),
+                        topics2.clone(),
+                        addresses.clone(),
+                        RETRY_LIMIT,
+                    )
+                    .await?;
+                let mut second_half = self
+                    .get_events_inner(
+                        BlockNumber::Number(mid + 1u64),
+                        to,
+                        topics1,
+                        topics2,
+                        addresses,
+                        RETRY_LIMIT,
+                    )
+                    .await?;
+
+                first_half.append(&mut second_half);
+                result = Ok(first_half);
+            } else if should_retry(err_code, err_message) && retries_left > 0 {
+                tracing::warn!("Retrying. Retries left: {retries_left}");
+                result = self
+                    .get_events_inner(from, to, topics1, topics2, addresses, retries_left - 1)
```

Comment:
nit: it seems wasteful to use recursion for retries instead of a loop, no?
Also the layout/block nesting of this function could be improved, by inverting success and failure branches, i.e.:
```
let err = match result {
  Ok(res) => return res;
  Err(err) => err
};
// error handling.
```